Normalise whitespace in testcases

kgaughan opened the pull request:
<!--
Thank you for your pull request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests and possibly benchmarks.

Contributors guide: ./CONTRIBUTING.md
-->

<!-- _Please make sure to review and check all of these items:_ -->


##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make test-all` (UNIX) passes. CI will also test this
- [x] unit and/or integration tests are included (if applicable)
- [x] documentation is changed or added (if applicable)

<!-- _NOTE: these things are not required to open a PR and can be done afterwards / while the PR is open._ -->

### Description of change
<!--
Please provide a description of the change here. Be sure to use issue references when applicable:
https://help.github.com/en/github/managing-your-work-on-github/closing-issues-using-keywords
-->

The .editorconfig file leads to trailing whitespace being removed. It's a bit of a coin-toss as to whether the right solution here is to override whitespace trimming for just the `testcase` directory or to instead perform some normalization of the expected and actual strings to trim this whitespace, so I did both.

kgaughan committed:
Update dependencies

Some of the messages the integration tests expect change in more recent
versions of Go.

Further, the .editorconfig file leads to trailing whitespace being
removed. It's a bit of a coin-toss as to whether the right solution here
is to override whitespace trimming for just the `testcase` directory or
to instead perform some normalization of the expected and actual strings
to trim this whitespace, so I did both.

kgaughan committed:
Update dependencies

Some of the messages the integration tests expect change in more recent
versions of Go.

Further, the .editorconfig file leads to trailing whitespace being
removed. It's a bit of a coin-toss as to whether the right solution here
is to override whitespace trimming for just the `testcase` directory or
to instead perform some normalization of the expected and actual strings
to trim this whitespace, so I did both.

petemounce commented:
This lgtm, but Travis must be persuaded to run. @aelsabbahy / @ripienaar - is that something you can address?

kgaughan committed:
Bump all dependencies that don't update the runtime

I've also fixed two spots that cause 'make test' to fail with a
'non-constant format string' error.

kgaughan commented:
@aelsabbahy Any chance you could spare some time to take a look at this? My most recent push to this PR addresses these issues associated with CVEs flagged by [govulncheck](https://go.dev/blog/govulncheck):

```console
$ govulncheck -show verbose ./...
Fetching vulnerabilities from the database...

Checking the code against the vulnerabilities...

The package pattern matched the following 7 root packages:
  github.com/goss-org/goss/matchers
  github.com/goss-org/goss/util
  github.com/goss-org/goss/system
  github.com/goss-org/goss/resource
  github.com/goss-org/goss/outputs
  github.com/goss-org/goss
  github.com/goss-org/goss/cmd/goss
Govulncheck scanned the following 46 modules and the go1.26.2 standard library:
  github.com/goss-org/goss
  dario.cat/mergo@v1.0.1
  github.com/Masterminds/goutils@v1.1.1
  github.com/Masterminds/semver/v3@v3.3.0
  github.com/Masterminds/sprig/v3@v3.3.0
  github.com/achanda/go-sysctl@v0.0.0-20160222034550-6be7678c45d2
  github.com/beorn7/perks@v1.0.1
  github.com/blang/semver/v4@v4.0.0
  github.com/cespare/xxhash/v2@v2.3.0
  github.com/cpuguy83/go-md2man/v2@v2.0.3
  github.com/fatih/color@v1.17.0
  github.com/google/go-cmp@v0.6.0
  github.com/google/uuid@v1.6.0
  github.com/goss-org/GOnetstat@v0.0.0-20230101144325-22be0bd9e64d
  github.com/goss-org/go-ps@v0.0.0-20230609005227-7b318e6a56e5
  github.com/hashicorp/logutils@v1.0.0
  github.com/huandu/xstrings@v1.5.0
  github.com/mattn/go-colorable@v0.1.13
  github.com/mattn/go-isatty@v0.0.20
  github.com/miekg/dns@v1.1.61
  github.com/mitchellh/copystructure@v1.2.0
  github.com/mitchellh/reflectwalk@v1.0.2
  github.com/moby/sys/mountinfo@v0.7.1
  github.com/munnerz/goautoneg@v0.0.0-20191010083416-a7dc8b61c822
  github.com/oleiade/reflections@v1.0.1
  github.com/onsi/gomega@v1.33.1
  github.com/patrickmn/go-cache@v2.1.0+incompatible
  github.com/pmezard/go-difflib@v1.0.0
  github.com/prometheus/client_golang@v1.19.1
  github.com/prometheus/client_model@v0.6.1
  github.com/prometheus/common@v0.55.0
  github.com/prometheus/procfs@v0.15.1
  github.com/russross/blackfriday/v2@v2.1.0
  github.com/samber/lo@v1.46.0
  github.com/shopspring/decimal@v1.4.0
  github.com/spf13/cast@v1.7.0
  github.com/tidwall/gjson@v1.17.1
  github.com/tidwall/match@v1.1.1
  github.com/tidwall/pretty@v1.2.1
  github.com/urfave/cli@v1.22.14
  golang.org/x/crypto@v0.26.0
  golang.org/x/net@v0.27.0
  golang.org/x/sys@v0.23.0
  golang.org/x/text@v0.17.0
  google.golang.org/protobuf@v1.34.2
  gopkg.in/yaml.v3@v3.0.1

=== Symbol Results ===

No vulnerabilities found.

=== Package Results ===

Vulnerability #1: GO-2026-4441
    Infinite parsing loop in golang.org/x/net
  More info: https://pkg.go.dev/vuln/GO-2026-4441
  Module: golang.org/x/net
    Found in: golang.org/x/net@v0.27.0
    Fixed in: golang.org/x/net@v0.45.0

Vulnerability #2: GO-2026-4440
    Quadratic parsing complexity in golang.org/x/net/html
  More info: https://pkg.go.dev/vuln/GO-2026-4440
  Module: golang.org/x/net
    Found in: golang.org/x/net@v0.27.0
    Fixed in: golang.org/x/net@v0.45.0

Vulnerability #3: GO-2025-3595
    Incorrect Neutralization of Input During Web Page Generation in x/net in
    golang.org/x/net
  More info: https://pkg.go.dev/vuln/GO-2025-3595
  Module: golang.org/x/net
    Found in: golang.org/x/net@v0.27.0
    Fixed in: golang.org/x/net@v0.38.0

Vulnerability #4: GO-2024-3333
    Non-linear parsing of case-insensitive content in golang.org/x/net/html
  More info: https://pkg.go.dev/vuln/GO-2024-3333
  Module: golang.org/x/net
    Found in: golang.org/x/net@v0.27.0
    Fixed in: golang.org/x/net@v0.33.0

=== Module Results ===

Vulnerability #1: GO-2025-4135
    Malformed constraint may cause denial of service in
    golang.org/x/crypto/ssh/agent
  More info: https://pkg.go.dev/vuln/GO-2025-4135
  Module: golang.org/x/crypto
    Found in: golang.org/x/crypto@v0.26.0
    Fixed in: golang.org/x/crypto@v0.45.0

Vulnerability #2: GO-2025-4134
    Unbounded memory consumption in golang.org/x/crypto/ssh
  More info: https://pkg.go.dev/vuln/GO-2025-4134
  Module: golang.org/x/crypto
    Found in: golang.org/x/crypto@v0.26.0
    Fixed in: golang.org/x/crypto@v0.45.0

Vulnerability #3: GO-2025-4116
    Potential denial of service in golang.org/x/crypto/ssh/agent
  More info: https://pkg.go.dev/vuln/GO-2025-4116
  Module: golang.org/x/crypto
    Found in: golang.org/x/crypto@v0.26.0
    Fixed in: golang.org/x/crypto@v0.43.0

Vulnerability #4: GO-2025-3503
    HTTP Proxy bypass using IPv6 Zone IDs in golang.org/x/net
  More info: https://pkg.go.dev/vuln/GO-2025-3503
  Module: golang.org/x/net
    Found in: golang.org/x/net@v0.27.0
    Fixed in: golang.org/x/net@v0.36.0

Vulnerability #5: GO-2025-3487
    Potential denial of service in golang.org/x/crypto
  More info: https://pkg.go.dev/vuln/GO-2025-3487
  Module: golang.org/x/crypto
    Found in: golang.org/x/crypto@v0.26.0
    Fixed in: golang.org/x/crypto@v0.35.0

Vulnerability #6: GO-2024-3321
    Misuse of connection.serverAuthenticate may cause authorization bypass in
    golang.org/x/crypto
  More info: https://pkg.go.dev/vuln/GO-2024-3321
  Module: golang.org/x/crypto
    Found in: golang.org/x/crypto@v0.26.0
    Fixed in: golang.org/x/crypto@v0.31.0
```

Now, I'll note that it also says this at the end:

```
Your code is affected by 0 vulnerabilities.
This scan also found 4 vulnerabilities in packages you import and 6
vulnerabilities in modules you require, but your code doesn't appear to call
these vulnerabilities.
```

So goss itself shouldn't be affected by any vulnerabilities, but it's still good hygiene to get them updated.

You might notice that I haven't updated `github.com/fatih/color` to the most recent version ([v1.19.0](https://github.com/fatih/color/releases/tag/v1.19.0)), as that would bump the runtime to 1.25, and I'd like to keep this simple.

aelsabbahy commented on matcher_test.go:
```diff
@@ -71,7 +71,7 @@ func TestMatchers(t *testing.T) {
 				t.Fatal(err)
 			}
-			wantOut := string(wantOutB)
+			wantOut := sanitizeTrailingWhitespace(string(wantOutB))
```
How does something like this compare?

```
strings.TrimRightFunc(s, unicode.IsSpace)
```

kgaughan replied:
I don't think that'd work as it'd just trim the end of the string rather than the ends of each line in the file.

kgaughan replied:
Yep, I've checked and it only trims the very end of each file.

aelsabbahy replied:
Oh, I see, it's a multi-line string.

Just noticed this is part of the test file, is goss adding training whitespace in its output?

Looking back at this, I'm thinking maybe the test should care about whitespace and be strict about expecting there to be no trialing whitespace.

That way if there's any regression where goss starts adding trialing whitespace, the test will fail.

aelsabbahy replied:
As in, duration is expected to be different unless we hack time, but formatting should be consistent.

kgaughan replied:
My main worry is that it's a bit of a trap for anyone who doesn't know to run `make update-matcher-tests`.

That said, I'm not too bothered about this getting merged; it's just that this is something I tripped up on myself initially.

aelsabbahy replied:
I'll merge it for now. The ideal is that goss doesn't print outputs with trailing whitespace.
